keps/sig-node/4381-dra-structured-parameters/README.md

```diff
@@ -2049,6 +2043,56 @@ Unreserve is called in two scenarios:
 
 ### kubelet
 
+#### Communication between kubelet and resource kubelet plugin
+
+Resource kubelet plugins are discovered through the [kubelet plugin registration
+mechanism](https://kubernetes.io/docs/concepts/extend-kubernetes/compute-storage-net/device-plugins/#device-plugin-registration). A
+new "ResourcePlugin" type will be used in the Type field of the
+[PluginInfo](https://pkg.go.dev/k8s.io/kubelet/pkg/apis/pluginregistration/v1#PluginInfo)
+response to distinguish the plugin from device and CSI plugins.
+
+Under the advertised Unix Domain socket the kubelet plugin provides the
+k8s.io/kubelet/pkg/apis/dra gRPC interface. It was inspired by
+[CSI](https://github.com/container-storage-interface/spec/blob/master/spec.md),
+with “volume” replaced by “resource” and volume specific parts removed.
+
+#### Version skew
+
+Previously, kubelet retrieved ResourceClaims and published ResourceSlices on
+behalf of DRA drivers on the node. The information included in those got passed
+between API server, kubelet, and kubelet plugin using the version of the
+resource.k8s.io used by the kubelet. Combining a kubelet using some older API
+version with a plugin using a new version was not possible because conversion
+of the resource.k8s.io types is only supported in the API server and an old
+kubelet wouldn't know about a new version anyway.
+
+Keeping kubelet at some old release while upgrading the control and DRA drivers
+is desirable and officially supported by Kubernetes. To support the same when
+using DRA, the kubelet now leaves ResourceSlice handling (almost) entirely to
+the plugins. The one exception is that it deletes all ResourceSlices on
+startup. This ensures that no pods depending in DRA get scheduled to the node
```

Comment:
On kubelet startup. I've discussed this a fair bit with @klueska (who knows the kubelet side of things much better than I do) and this is what we settled on.

The rationale (@klueska, please correct me if I got something wrong) is that:
- The kubelet cannot really distinguish between node start/node reboot/kubelet restart.
- Driver registration might take an arbitrarily long time, during which there is incorrect resource information for the node. This might cause pods to get scheduled which then cannot run.

We decided to play it safe and discard existing objects. If/when drivers are running, the information gets restored, which is also when pods can get scheduled again.

My biggest concern was the extra traffic that this causes. It would have been nice to reuse existing objects, but I couldn't find a good heuristic when to delete and when to keep.

Reply:
I'm really surprised resourceslices can just be deleted with no adverse effects to already scheduled / running pods

Reply:
Currently, allocation is tracked in the ResourceClaim, not the ResourceSlice.  Claims don't reference specific ResourceSlice objects. They reference devices with "driver name/node name/device name", so that reference survives recreation of the ResourceSlice objects.

A driver needs to check that it still has the device referenced in a claim, but it doesn't need the ResourceSlice for that because it's the authority about what is and what isn't available and usable.

Reply:
Added to the KEP.